egs/wsj/s5/steps/nnet3/train_rnn.py

```diff
@@ -0,0 +1,1132 @@
+        
+        if iter in realign_iters:
+            prev_egs_dir=cur_egs_dir
+            cur_egs_dir="{0}/egs_{1}".format(args.dir, "iter"+str(iter))
+            Realign(args.dir, iter, args.alidir, args.lang,
+                    prev_egs_dir, cur_egs_dir,
+                    args.prior_subset_size, num_archives, run_opts,
+                    transform_dir = args.transform_dir, online_ivector_dir = args.online_ivector_dir)
+            if args.cleanup and args.egs_dir is None:
+                RemoveEgs(prev_egs_dir)
+
+        shrinkage_value = args.shrink_value if DoShrinkage(iter, model_file, "SigmoidComponent", args.shrink_threshold) else 1
+        logger.info("On iteration {0}, learning rate is {1} and shrink value is {2}.".format(iter, learning_rate(iter), shrinkage_value))
+
+        if args.stage <= iter:
+            TrainOneIteration(args.dir, iter, egs_dir,
+                              learning_rate, num_chunk_per_minibatch,
+                              num_hidden_layers, add_layers_period,
+                              run_opts)
+        if args.cleanup:
+            # do a clean up of the previous model
+            CleanNnetDir(args.dir, num_iters, egs_dir,
+                 preserve_model_interval = preserve_model_interval,
+                 remove_egs = False)
+
+        num_archives_processed = num_archives_processed + cur_num_jobs
+
+    if args.stage <= num_iters:
+        logger.info("Doing final combination to produce final.mdl")
+        CombineModels(args.dir, num_iters, num_iters_combine, args.chunk_width,
+                      egs_dir, run_opts)
+
+    if args.stage <= num_iters + 1:
+        logger.info("Getting average posterior for purposes of adjusting the priors.")
+        avg_post_vec_file = ComputeAveragePosterior(dir, 'combined', egs_dir,
+                                num_archives, prior_subset_size, run_opts)
+
+    logger.info("Re-adjusting priors based on computed posteriors")
+    combined_model = "{dir}/combined.mdl".format(dir = dir)
+    final_model = "{dir}/final.mdl".format(dir = args.dir)
+    AdjustAmPriors(dir, combined_model, avg_post_vec_file, final_model)
+
+    if args.clean_up:
+        logger.info("Cleaning up the experiment directory {0}".format(args.dir))
+        remove_egs = args.remove_egs
+        if args.egs_dir is not None:
+            # this egs_dir was not created by this experiment so we will not
+            # delete it
+            remove_egs = False
+
+        CleanNnetDir(args.dir, num_iters, cur_egs_dir,
+                     preserve_model_interval = args.preserve_model_interval,
+                     remove_egs = remove_egs)
+
+# a class to store run options
+class RunOpts:
+    command
+    train_queue_opt
+    combine_queue_opt
+    prior_gpu_opt
+    prior_queue_opt
+    parallel_train_opts
+    
+    def __init__():
+        pass
+
+
+if __name__ == "__main__":
```

Comment:
Just a couple, mostly stylistic/cosmetic, suggestions. I'm by no means a Python guru, so I don't claim that what I propose is necessarily better than the current code..

Alternatively you might use:

```
def main():
  <actual top-level code here>

if __name__ == '__main__':
  main()
```

That way the variables in the body of main() will be local. As it is now, *args* and *run_opts*, for example, are in fact visible in all functions. This is usually is not a problem in my experience, but once in a while may cause some head-scratching.